README
Sample readme file for at2 project.


NEWS
Sample NEWS file for at2 project.


src/test.c
/*
 * test.c
 *
 *  Created on: 08 04 2015
 *      Author: administrator
 *
 *  Notebook edition
 */

// BUILD IT WITH -O0 and -msse options to see most difference
//
// benchmark array size (floats) here!
#define FLOAT_ARRAY_SIZE 96000000

#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include "test.h"

int main()
{
	struct timeval timeStart, timeEnd;
	struct timeval timeFloatAddStart, timeVectorAddStart;
	struct timeval timeFloatAddEnd, timeVectorAddEnd;
	double elapsedTime;

	gettimeofday(&timeStart, NULL);

	//	two input arrays and one output (A + B = C)
	//	that`s test =)
	float *fA = (float *) malloc(sizeof(float) * FLOAT_ARRAY_SIZE);
	float *fB = (float *) malloc(sizeof(float) * FLOAT_ARRAY_SIZE);
	float *fC = (float *) malloc(sizeof(float) * FLOAT_ARRAY_SIZE);

	// init input arrays with some rubbish
	// INIT_GROW is +1 sequence of values start with valueA for vA (this start = 111) and valueB for vB (this start = 222)
	// so first value in vC shall be 333.
	// INIT_EASY is just a fill with valueA
	initFloatAB(fA, fB, INIT_GROW, 111, 222);

	// A + B = C with default float arrays
	gettimeofday(&timeFloatAddStart, NULL);
	floatArrayAdd(fA, fB, fC);
	gettimeofday(&timeFloatAddEnd, NULL);

	// two input arrays and one output (A + B = C)
	// with using type of vecm, so tests must be done with
	// variation of compilers and optimization flags to catch and see the difference
	vecm *vA = (vecm *) malloc(sizeof(vecm) * VECTOR_ARRAY_SIZE);
	vecm *vB = (vecm *) malloc(sizeof(vecm) * VECTOR_ARRAY_SIZE);
	vecm *vC = (vecm *) malloc(sizeof(vecm) * VECTOR_ARRAY_SIZE);

	// init input arrays with some rubbish
	// INIT_GROW is +1 sequence of values start with valueA for vA (this start = 111) and valueB for vB (this start = 222)
	// so first value in vC shall be 333.
	// INIT_EASY is just a fill with valueA
	initVectorFloatAB(vA, vB, INIT_GROW, 111, 222);

	// A + B = C with vector array and vector type operation
	gettimeofday(&timeVectorAddStart, NULL);
	floatVectorArrayAdd(vA, vB, vC);
	gettimeofday(&timeVectorAddEnd, NULL);

	//fprintresult(fA, fB, fC);
	//vprintresult(vA, vB, vC);

	printf("Bench addition of two %d-element arrays of float:\n", FLOAT_ARRAY_SIZE);

	//printFloatArrays(fA,fB,fC);
	//printVectorArrays(vA,vB,vC);

	// free memory (compatibilty with some old compilers)
	free(fA);
	free(fB);
	free(fC);
	free(vA);
	free(vB);
	free(vC);

	gettimeofday(&timeEnd, NULL);

	elapsedTime = (timeEnd.tv_sec - timeStart.tv_sec) * 1000.0;      // sec to ms
	elapsedTime += (timeEnd.tv_usec - timeStart.tv_usec) / 1000.0;   // us to ms

	printf("\nTotal time with memory allocation/free is: %.2f ms\n",elapsedTime);

	elapsedTime = (timeFloatAddEnd.tv_sec - timeFloatAddStart.tv_sec) * 1000.0;      // sec to ms
	elapsedTime += (timeFloatAddEnd.tv_usec - timeFloatAddStart.tv_usec) / 1000.0;   // us to ms

	printf("\nAddition of two arrays of float[%d]: %.2f ms\n",FLOAT_ARRAY_SIZE,elapsedTime);

	elapsedTime = (timeVectorAddEnd.tv_sec - timeVectorAddStart.tv_sec) * 1000.0;      // sec to ms
	elapsedTime += (timeVectorAddEnd.tv_usec - timeVectorAddStart.tv_usec) / 1000.0;   // us to ms

	printf("\nAddition of two arrays of vector float[%d]: %.2f ms\n",FLOAT_ARRAY_SIZE,elapsedTime);

	return (0);
}

void initFloatAB(float *fA, float *fB, int mode, int valueA, int valueB)
{
	switch (mode)
	{
	case INIT_EASY:
		initFloatArray(fA, INT, valueA, ZERO);
		initFloatArray(fB, INT, valueB, ZERO);
		break;
	case INIT_GROW:
		initFloatArray(fA, GROW, valueA, ZERO);
		initFloatArray(fB, GROW, valueB, ZERO);
		break;
	default:
		break;
	}
}

void initVectorFloatAB(vecm *vA, vecm *vB, int mode, int valueA, int valueB)
{
	switch (mode)
	{
	case INIT_EASY:
		initVectorArray(vA, INT, valueA, ZERO);
		initVectorArray(vB, INT, valueB, ZERO);
		break;
	case INIT_GROW:
		initVectorArray(vA, GROW, valueA, valueB);
		initVectorArray(vB, GROW, valueB, valueA);
		break;
	default:
		break;
	}
}

// valueB is not used. future feature stub
void initFloatArray(float *fA, int initMode, int valueA, int valueB)
{
	int i;
	switch (initMode)
	{
	case INT:
		for (i = 0; i < FLOAT_ARRAY_SIZE; i++)
		{
			fA[i] = (float) valueA;
		}
		break;
	case GROW:
		for (i = 0; i < FLOAT_ARRAY_SIZE; i++)
		{
			fA[i] = (float)valueA++ / 3.14159;
		}
		break;
	default:
		break;
	}
}

// valueB is not used. future feature stub
void initVectorArray(vecm *vA, int initMode, int valueA, int valueB)
{
	int i, j;
	float *vfA;
	switch (initMode)
	{
	case INT:
		for (i = 0; i < VECTOR_ARRAY_SIZE; i++)
		{
			vfA = (float *) &vA[i];
			for (j = 0; j < VECTORSIZE; j++)
			{
				vfA[j] = valueA;
			}
		}
		break;
	case GROW:
		for (i = 0; i < VECTOR_ARRAY_SIZE; i++)
		{
			vfA = (float *) &vA[i];
			for (j = 0; j < VECTORSIZE; j++)
			{
				vfA[j] = valueA++ / 3.14159;
			}
		}
		break;
	default:
		break;
	}
}

void floatArrayAdd(float *fA, float *fB, float *fC)
{
	int i;
	for (i = 0; i < FLOAT_ARRAY_SIZE; i++)
	{
		fC[i] = fA[i] + fB[i];
	}
}

void floatVectorArrayAdd(vecm *vA, vecm *vB, vecm *vC)
{
	int i;
	for (i = 0; i < VECTOR_ARRAY_SIZE; i++)
	{
		vC[i] = vA[i] + vB[i];
	}
}

void printFloatArrays(float *fA, float *fB, float *fC)
{
	int i;
	for (i = 0; i < FLOAT_ARRAY_SIZE; i++)
	{
		printf("farray #%d: (A)%f + (B)%f = (C)%f\n", i, fA[i], fB[i], fC[i]);
	}
}

void printVectorArrays(vecm *vA, vecm *vB, vecm *vC)
{
	int i, j;
	float *vfA;
	float *vfB;
	float *vfC;
	for (i = 0; i < VECTOR_ARRAY_SIZE; i++)
	{
		vfA = (float *) &vA[i];
		vfB = (float *) &vB[i];
		vfC = (float *) &vC[i];
		for (j = 0; j < VECTORSIZE; j++)
		{
			printf("vector array element [#%d]: (A) %f + (B) %f = (C) %f\n", i * VECTORSIZE + j,
					vfA[j], vfB[j], vfC[j]);
		}
	}
}




src/src/test.c
/*
 * test.c
 *
 *  Created on: 08 04 2015
 *      Author: administrator
 *
 *	Server edition (for 64 * 1000 000 elements you need ~500 Mb of memory)
 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include "test.h"

int main() {

	int i,n;

	double elapsedTime;

	struct timeval timeStart, timeEnd;
	struct timeval timeFloatAddStart, timeVectorAddStart;
	struct timeval timeFloatAddEnd, timeVectorAddEnd;
	struct timeval timeFloatInitStart, timeVectorInitStart;
	struct timeval timeFloatInitEnd, timeVectorInitEnd;

	gettimeofday(&timeStart, NULL);

	//	two input arrays and one output (A + B = C)
	//	that`s test =)
	float *fA = (float *) malloc(sizeof(float) * FLOAT_ARRAY_SIZE);
	float *fB = (float *) malloc(sizeof(float) * FLOAT_ARRAY_SIZE);
	float *fC = (float *) malloc(sizeof(float) * FLOAT_ARRAY_SIZE);

	// init input arrays with some rubbish
	// INIT_GROW is +1 sequence of values start with valueA for vA (this start = 111) and valueB for vB (this start = 222)
	// so first value in vC shall be 333.
	// INIT_EASY is just a fill with valueA
	gettimeofday(&timeFloatInitStart, NULL);
	initFloatAB(fA, fB, INIT_GROW, 111, 222);
	gettimeofday(&timeFloatInitEnd, NULL);

	// A + B = C with default float arrays
	gettimeofday(&timeFloatAddStart, NULL);
	floatArrayAdd(fA, fB, fC);
	gettimeofday(&timeFloatAddEnd, NULL);

	// two input arrays and one output (A + B = C)
	// with using type of vecm, so tests must be done with
	// variation of compilers and optimization flags to catch and see the difference
	vecm *vA = (vecm *) malloc(sizeof(vecm) * VECTOR_ARRAY_SIZE);
	vecm *vB = (vecm *) malloc(sizeof(vecm) * VECTOR_ARRAY_SIZE);
	vecm *vC = (vecm *) malloc(sizeof(vecm) * VECTOR_ARRAY_SIZE);

	// init input arrays with some rubbish
	// INIT_GROW is +1 sequence of values start with valueA for vA (this start = 111) and valueB for vB (this start = 222)
	// so first value in vC shall be 333.
	// INIT_EASY is just a fill with valueA
	gettimeofday(&timeVectorInitStart, NULL);
	initVectorFloatAB(vA, vB, INIT_GROW, 111, 222);
	gettimeofday(&timeVectorInitEnd, NULL);

	// A + B = C with vector array and vector type operation
	gettimeofday(&timeVectorAddStart, NULL);
	floatVectorArrayAdd(vA, vB, vC);
	gettimeofday(&timeVectorAddEnd, NULL);

	//fprintresult(fA, fB, fC);
	//vprintresult(vA, vB, vC);

	printf("Bench addition of two %d-element arrays of float:\n",
			FLOAT_ARRAY_SIZE);

	printFloatArrays(fA,fB,fC,DEBUG_LINES);
	printf("\n\n");
	printVectorArrays(vA,vB,vC,DEBUG_LINES);

	// free memory (compatibilty with some old compilers)
	free(fA);
	free(fB);
	free(fC);
	free(vA);
	free(vB);
	free(vC);

	gettimeofday(&timeEnd, NULL);

	elapsedTime = (timeEnd.tv_sec - timeStart.tv_sec) * 1000.0;     // sec to ms
	elapsedTime += (timeEnd.tv_usec - timeStart.tv_usec) / 1000.0;   // us to ms

	printf("\nTotal time with memory allocation/free is: %.2f ms\n",
			elapsedTime);

	elapsedTime = (timeFloatInitEnd.tv_sec - timeFloatInitStart.tv_sec) * 1000.0; // sec to ms
	elapsedTime += (timeFloatInitEnd.tv_usec - timeFloatInitStart.tv_usec)
			/ 1000.0;   // us to ms

	printf("\nInit of two arrays of float[%d]: %.2f ms\n", FLOAT_ARRAY_SIZE,
			elapsedTime);

	elapsedTime = (timeVectorInitEnd.tv_sec - timeVectorInitStart.tv_sec)
			* 1000.0;      // sec to ms
	elapsedTime += (timeVectorInitEnd.tv_usec - timeVectorInitStart.tv_usec)
			/ 1000.0;   // us to ms

	printf("\nInit of two arrays of vector float[%d]: %.2f ms\n",
			FLOAT_ARRAY_SIZE, elapsedTime);

	elapsedTime = (timeFloatAddEnd.tv_sec - timeFloatAddStart.tv_sec) * 1000.0; // sec to ms
	elapsedTime += (timeFloatAddEnd.tv_usec - timeFloatAddStart.tv_usec)
			/ 1000.0;   // us to ms

	printf("\nAddition of two arrays of float[%d]: %.2f ms\n", FLOAT_ARRAY_SIZE,
			elapsedTime);

	elapsedTime = (timeVectorAddEnd.tv_sec - timeVectorAddStart.tv_sec)
			* 1000.0;      // sec to ms
	elapsedTime += (timeVectorAddEnd.tv_usec - timeVectorAddStart.tv_usec)
			/ 1000.0;   // us to ms

	printf("\nAddition of two arrays of vector float[%d]: %.2f ms\n",
			FLOAT_ARRAY_SIZE, elapsedTime);

	return (0);
}

void floatArrayAdd(float *fA, float *fB, float *fC) {
	int i;
	for (i = 0; i < FLOAT_ARRAY_SIZE; i++) {
		fC[i] = fA[i] + fB[i];
	}
}

void floatVectorArrayAdd(vecm *vA, vecm *vB, vecm *vC) {
	int i;
	for (i = 0; i < VECTOR_ARRAY_SIZE; i++) {
		vC[i] = vA[i] + vB[i];
	}
}

void printFloatArrays(float *fA, float *fB, float *fC, int lines) {
	int i;
	for (i = 0; i < lines; i++) {
		printf("float array element [#%d]: (A) %f + (B) %f = (C) %f\n", i, fA[i], fB[i], fC[i]);
	}
}

void printVectorArrays(vecm *vA, vecm *vB, vecm *vC, int lines) {
	int i, j;
	float *vfA;
	float *vfB;
	float *vfC;
	for (i = 0; i < lines>>2; i++) {
		vfA = (float *) &vA[i];
		vfB = (float *) &vB[i];
		vfC = (float *) &vC[i];
		for (j = 0; j < VECTORSIZE; j++) {
			printf("vector array element [#%d]: (A) %f + (B) %f = (C) %f\n",
					i * VECTORSIZE + j, vfA[j], vfB[j], vfC[j]);
		}
	}
}



src/src/init.c
#include <def.h>

#ifdef OPENMP_ENABLE
//#define OPENMP
#endif

// valueB is not used. future feature stub
void initFloatArray(float *fA, int initMode, int valueA, int valueB) {

	int i;
	float n;

	switch (initMode) {
	case INT:
		for (i = 0; i < FLOAT_ARRAY_SIZE; i++) {
			fA[i] = (float) valueA;
		}
		break;
	case GROW:

#ifdef OPENMP
#pragma omp parallel shared(fA) private(i,n)
{
		n = (float)(omp_get_thread_num()+1.1)/10 + valueA;

		#pragma omp for
		for (i = 0; i < FLOAT_ARRAY_SIZE; i++) {
			fA[i] = i+n;
		}
}
#else
		n = valueA;
		for (i = 0; i < FLOAT_ARRAY_SIZE; i++) {
			fA[i] = i+n;
		}
#endif

		break;
	default:
	break;
	}
}

// valueB is not used. future feature stub
void initVectorArray(vecm *vA, int initMode, int valueA, int valueB) {

	int i, j;
	float n;
	float *vfA;

	switch (initMode) {
	case INT:
		for (i = 0; i < VECTOR_ARRAY_SIZE; i++) {
			vfA = (float *) &vA[i];
			for (j = 0; j < VECTORSIZE; j++) {
				vfA[j] = valueA;
			}
		}
		break;
	case GROW:

#ifdef OPENMP
#pragma omp parallel shared(vA) private(i,n) /* add vfA into private vars to use slow pointer loop */
{
		// rational part as thread tag
		n = (float)(omp_get_thread_num()+1.1)/10;

		#pragma omp for
		for (i = 0; i < VECTOR_ARRAY_SIZE; i++) {

			// slower
			//vA[i] = (vecm){i<<2,(i<<2)+1,(i<<2)+2,(i<<2)+3} + (vecm){n,n,n,n};

			// faster
			vA[i] = (vecm){(i<<2)+n,(i<<2)+1+n,(i<<2)+2+n,(i<<2)+3+n};
		}
}
#else
		n = 0.777777;
		for (i = 0; i < VECTOR_ARRAY_SIZE; i++) {
			// slower
			//vA[i] = (vecm){i<<2,(i<<2)+1,(i<<2)+2,(i<<2)+3} + (vecm){n,n,n,n};

			// faster
			vA[i] = (vecm){(i<<2)+n,(i<<2)+1+n,(i<<2)+2+n,(i<<2)+3+n};
		}
#endif

		break;
	default:
		break;
	}
}


src/include/def.h
/*
 * test.h
 *
 *  Created on: 09 04 2015
 *      Author: administrator
 */

#ifndef TEST_H_
#define TEST_H_

#include <omp.h>

/* benchmark array size (floats) here! */
#define FLOAT_ARRAY_SIZE 100000000
#define DEBUG_LINES 4
#define BENCH_PASSES 10
/*-------------------------------------*/

#define ZERO 0

#define INIT_EASY 1
#define INIT_GROW 2

#define INT 1
#define GROW 2

#define VECTORSIZE 4 /* does it going right? */

typedef float __attribute__ ((vector_size (sizeof(float) * VECTORSIZE))) vecm;

#define VECTOR_ARRAY_SIZE FLOAT_ARRAY_SIZE/VECTORSIZE

void initFloatArraySIMD(float *, int, int, int);
void initFloatArraySIMDMP(float *, int, int, int);

void initFloatArray(float *, int, int, int);
void initVectorArray(vecm *, int, int, int);

void initFloatAB(float *, float *, int, int, int);
void initVectorFloatAB(vecm *, vecm *, int, int, int);

void floatArrayAdd(float *, float *, float *);
void floatVectorArrayAdd(vecm *, vecm *, vecm *);

void printFloatArrays(float *, float *, float *, int);
void printVectorArrays(vecm *, vecm *, vecm *,int);

#include <sys/time.h>

typedef void (*pF)(float *,float *,int,int,int);
void benchIT(struct timeval*,struct timeval*, pF );

#endif /* TEST_H_ */


src/src/vectorbench.c
/*
 * test.c
 *
 *  Created on: 08 04 2015
 *      Author: administrator
 *
 *	Server edition (for 64 * 1000 000 elements you need ~500 Mb of memory)
 */

#include <stdio.h>
#include <stdlib.h>
#include "def.h"

int main(int argc,char **argv) {

	int b,i,n;
	double elapsedTime;

	struct timeval timeStart, timeEnd;
	struct timeval timeFloatAddStart, timeVectorAddStart;
	struct timeval timeFloatAddEnd, timeVectorAddEnd;
	struct timeval timeFloatInitStart, timeVectorInitStart;
	struct timeval timeFloatInitEnd, timeVectorInitEnd;

	// debug only
	setvbuf(stdout, NULL, _IONBF, 0);

	// start time profile
	gettimeofday(&timeStart, NULL);

	//	two input arrays and one output (A + B = C)
	//	that`s test =)
	float *fA = (float *) malloc(sizeof(float) * FLOAT_ARRAY_SIZE);
	if (!fA) {
		printf("malloc A failed\n");
		return(1);
	}

	float *fB = (float *) malloc(sizeof(float) * FLOAT_ARRAY_SIZE);
	if (!fB) {
		printf("malloc B failed\n");
		return(1);
	}

	float *fC = (float *) malloc(sizeof(float) * FLOAT_ARRAY_SIZE);
	if (!fC) {
		printf("malloc C failed\n");
		return(1);
	}

	printf ("Mem assigned for float arrays: %d bytes\n",FLOAT_ARRAY_SIZE * 3 * sizeof(fA));

	// init input arrays with some rubbish
	// INIT_GROW is +1 sequence of values start with valueA for vA (this start = 111) and valueB for vB (this start = 222)
	// so first value in vC shall be 333.
	// INIT_EASY is just a fill with valueA
	gettimeofday(&timeFloatInitStart, NULL);
	for (b=0;b<BENCH_PASSES;b++) {
		initFloatAB(fA, fB, INIT_GROW, 111, 222);
		printf("initFloat pass %d done...\n",b);
	}
	gettimeofday(&timeFloatInitEnd, NULL);

	//benchIT(&timeFloatInitStart,&timeFloatInitEnd, (*pF)(initFloatAB(fA, fB, INIT_GROW, 111, 222)) );

	// initFloatAB(fA, fB, INIT_GROW, 111, 222)

	// A + B = C with default float arrays
	gettimeofday(&timeFloatAddStart, NULL);
	floatArrayAdd(fA, fB, fC);
	gettimeofday(&timeFloatAddEnd, NULL);

	printf("Float arrays bench done...\n");

	// free memory (compatibilty with some old compilers)
	free(fA);
	free(fB);
	free(fC);

	// two input arrays and one output (A + B = C)
	// with using type of vecm, so tests must be done with
	// variation of compilers and optimization flags to catch and see the difference
	vecm *vA = (vecm *) malloc(sizeof(vecm) * VECTOR_ARRAY_SIZE);
	if (!vA) {
		printf("malloc vA failed\n");
		return(1);
	}

	vecm *vB = (vecm *) malloc(sizeof(vecm) * VECTOR_ARRAY_SIZE);
	if (!vB) {
		printf("malloc vB failed\n");
		return(1);
	}

	vecm *vC = (vecm *) malloc(sizeof(vecm) * VECTOR_ARRAY_SIZE);
	if (!vC) {
		printf("malloc vC failed\n");
		return(1);
	}

	printf ("Mem assigned for vector float arrays: %d bytes\n",VECTOR_ARRAY_SIZE * 3 * sizeof(*vA));

	// init input arrays with some rubbish
	// INIT_GROW is +1 sequence of values start with valueA for vA (this start = 111) and valueB for vB (this start = 222)
	// so first value in vC shall be 333.
	// INIT_EASY is just a fill with valueA
	gettimeofday(&timeVectorInitStart, NULL);
	initVectorFloatAB(vA, vB, INIT_GROW, 111, 222);
	gettimeofday(&timeVectorInitEnd, NULL);

	// A + B = C with vector array and vector type operation
	gettimeofday(&timeVectorAddStart, NULL);
	floatVectorArrayAdd(vA, vB, vC);
	gettimeofday(&timeVectorAddEnd, NULL);

	printf("Float vector arrays bench done...\n------------------------------------------------------------------------\n");

	//fprintresult(fA, fB, fC);
	//vprintresult(vA, vB, vC);

	printf("Bench addition of two %d-element arrays of float:\n",
			FLOAT_ARRAY_SIZE);

	printFloatArrays(fA,fB,fC,DEBUG_LINES);
	printf("\n\n");
	printVectorArrays(vA,vB,vC,DEBUG_LINES);

	// free memory (compatibilty with some old compilers)
	free(vA);
	free(vB);
	free(vC);

	gettimeofday(&timeEnd, NULL);

	elapsedTime = (timeEnd.tv_sec - timeStart.tv_sec) * 1000.0;     // sec to ms
	elapsedTime += (timeEnd.tv_usec - timeStart.tv_usec) / 1000.0;   // us to ms

	printf("\nTotal time with memory allocation/free is: %.2f ms\n",
			elapsedTime);

	elapsedTime = (timeFloatInitEnd.tv_sec - timeFloatInitStart.tv_sec) * 1000.0; // sec to ms
	elapsedTime += (timeFloatInitEnd.tv_usec - timeFloatInitStart.tv_usec) / 1000.0;   // us to ms

	printf("\nwieghted init of two arrays of float[%d]: ~%.2f ms\n", FLOAT_ARRAY_SIZE,
			elapsedTime/BENCH_PASSES);

	elapsedTime = (timeVectorInitEnd.tv_sec - timeVectorInitStart.tv_sec)
			* 1000.0;      // sec to ms
	elapsedTime += (timeVectorInitEnd.tv_usec - timeVectorInitStart.tv_usec)
			/ 1000.0;   // us to ms

	printf("\nwieghted init of two arrays of vector float[%d]: ~%.2f ms\n",
			FLOAT_ARRAY_SIZE/BENCH_PASSES, elapsedTime);

	elapsedTime = (timeFloatAddEnd.tv_sec - timeFloatAddStart.tv_sec) * 1000.0; // sec to ms
	elapsedTime += (timeFloatAddEnd.tv_usec - timeFloatAddStart.tv_usec)
			/ 1000.0;   // us to ms

	printf("\nAddition of two arrays of float[%d]: %.2f ms\n", FLOAT_ARRAY_SIZE,
			elapsedTime);

	elapsedTime = (timeVectorAddEnd.tv_sec - timeVectorAddStart.tv_sec)
			* 1000.0;      // sec to ms
	elapsedTime += (timeVectorAddEnd.tv_usec - timeVectorAddStart.tv_usec)
			/ 1000.0;   // us to ms

	printf("\nAddition of two arrays of vector float[%d]: %.2f ms\n",
			FLOAT_ARRAY_SIZE, elapsedTime);

	return (0);
}

void floatArrayAdd(float *fA, float *fB, float *fC) {
	int i;
	for (i = 0; i < FLOAT_ARRAY_SIZE; i++) {
		fC[i] = fA[i] + fB[i];
	}
}

void floatVectorArrayAdd(vecm *vA, vecm *vB, vecm *vC) {
	int i;
	for (i = 0; i < VECTOR_ARRAY_SIZE; i++) {
		vC[i] = vA[i] + vB[i];
	}
}

void printFloatArrays(float *fA, float *fB, float *fC, int lines) {
	int i;
	for (i = 0; i < lines; i++) {
		printf("float array element [#%d]: (A) %f + (B) %f = (C) %f\n", i, fA[i], fB[i], fC[i]);
	}
}

void printVectorArrays(vecm *vA, vecm *vB, vecm *vC, int lines) {
	int i, j;
	float *vfA;
	float *vfB;
	float *vfC;
	for (i = 0; i < lines>>2; i++) {
		vfA = (float *) &vA[i];
		vfB = (float *) &vB[i];
		vfC = (float *) &vC[i];
		for (j = 0; j < VECTORSIZE; j++) {
			printf("vector array element [#%d]: (A) %f + (B) %f = (C) %f\n",
					i * VECTORSIZE + j, vfA[j], vfB[j], vfC[j]);
		}
	}
}

/*void benchIT(struct timeval *timeStart,struct timeval *timeEnd, pF ) {
	int i = BENCH_PASSES;
	gettimeofday(timeStart, NULL);
	for (i=0;i<BENCH_PASSES;i++) {
		//f(fA, fB, INIT_GROW, 111, 222);
		printf("initFloat pass %d done...\n",i);
	}
	gettimeofday(timeEnd, NULL);
}*/
